hivemind/client/averaging/allreduce.py

```diff
@@ -59,6 +62,7 @@ async def accumulate_part(self, source: Endpoint, remote_part: torch.Tensor) ->
         assert not self.future.done(), f"already finished allreduce: {self.future}"
         assert source in self.local_tensor_parts, "unexpected source, not a part of current group"
         assert source not in self.accumulated_from, "duplicate source, already received that part"
+        assert not self.endpoint in self.client_mode_endpoints, f"{self.endpoint} is in client mode"
```

Comment:
If we want to disallow calling `accumulate_part` from a client-only participants should I add a separate test checking a failure?

Reply:
@AllreduceException - In future it would be great to introduce a separate exception for this. As of now, however, the AllreduceException is used for errors that are "normal", e.g. caused by a network failure.
In turn, errors that indicate hivemind implementation problems are raised as AssertionError/ValueError

@separate test - that would be great, but not absolutely required

Reply:
Shouldn't accumulate_part be simply a no-op for client-only participants? I guess the cleanest way to implement averaging would be to have the same execution flow for all workers regardless of their participation status. 

So ideally if you want to join the training process with zero bandwidth, all you have to do is set your bandwidth to 0 — the code remains the same. We may short-circuit that and avoid calling unnecessary functions under the hood, but from the user point of view the API has to be the same

Reply:
[based on our previous conversation]
Averager syntax should definitely be the same for the user, but accumulate_part is not exposed to the user and calling it will incur at least 1 RTT delay.

As a recommendation - let's keep the assertion in accumulate_part and make sure that the rest of our DecentralizedAverager respects it.